pkg/policy/rule.go

```diff
@@ -66,6 +71,30 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
+
+	if filter.addEndpoints(endpoints) && r.NumRules() == 0 {
+		// skip this policy as it is already covered and it does not contain L7 rules
```

Comment:
I think this is incorrect.

The `addEndpoints` method only resolves which endpoints are allowed to connect at L3 and above.
When `filter.addEndpoints` returns true, it only means that connectivity at L3 was already allowed. But it provides no indication whether there are L7 rules involved, i.e. L3/label-dependent L7.

If the new rule has no L7 rules, it means that the new rule is L3+L4 only. If there were L3+L4+L7 rules associated with this L4 port and endpoint selector, those rules must be replaced with the L3+L4-only rule. I.e. you must remove any existing L7 rule associated with the endpoint selectors.

Reply:
You're right, this does not check whether the current L4Filter has L7 rules involved. We need to perform similar logic to what I have below for cleaning out the L7-related policy.

Reply:
Fixed and added a unit tests.